### ISSUE TITLE
Make it possible to install Active Directory in the container (Windows Server 2022)

### ISSUE THREAD

**Issue by johlju** [microsoft/Windows-Containers#116]
We in the [DSC Community](dsccommunity.org) have many DSC resources that configures various aspect of the Windows eco-system (examples https://github.com/dsccommunity/). During the development and prior to deploy we like to run integration tests either in lab or in deploy pipelines. Many of the DSC modules requires Active Directory to function which currently hinders us from running integration tests since installing AD requires a restart of the build worker (which breaks the pipeline). Because of that there are many DSC resources that are released untested (other than unit testing). Some modules like the [ActiveDirectoryDsc](https://www.powershellgallery.com/packages/ActiveDirectoryDsc/6.2.0-preview0001) requires a lot of machines to fully integration test (forests, child domains, trusted domains etc.) it takes a pretty large lab system to run manually integration tests.

If it were possible to install Active Directory in the container (totally okay if it is unsupported for production) then we could use leverage several containers on the same host and set up and tear-down an environment per integration test or a group of integration tests. 

The benefit of this is that contributors of the DSC Community could get there contributions reviewed so much faster when we can fully automate the integration tests, in the already automated deploy pipeline.

I would not expect this functionality for older operating systems, but for Windows Server 2022 it would be natural for this being an option (again, would not expect it for production usage).

**Comment by vrapolinario** [comment]
Thanks for the detailed notes on the request. I think there are two things to look at here:
First, can you use a VM image that has AD pre-installed? I believe there are some considerations for generalizing an image that is a DC, but it should work.
Second, is the request for AD on Containers - which we've heard from customers already. For this specific scenario, do you need this to run on a DC or you need this to run on an instance that is domain joined?

**Comment by johlju** [comment]
> First, can you use a VM image that has AD pre-installed? I believe there are some considerations for generalizing an image that is a DC, but it should work.

In this case I do not want to have AD already installed in a VM. The goal of the PowerShell module ActiveDirectoryDsc is to install, configure and monitor drift using PowerShell Desired State Configuration (DSC) (using both Windows PowerShell and hopefully soon PowerShell too). So in this case I want to start in a fresh VM (or better yet a container) and using DSC with the module ActiveDirectoryDsc I want to integration test to install and configure Active Directory. This can be done using a VM in any hypervisor product today, but it takes a lot of VM's that need to be provisioned and deprovisioned for each new Pull Request.

> Second, is the request for AD on Containers - which we've heard from customers already. For this specific scenario, do you need this to run on a DC or you need this to run on an instance that is domain joined?

I this scenario I don't need the host/VM that is running the container do be either a DC or domain joined. Non-member server is enough. Though if the host that is running container is domain joined, it should be to a production Active Directory and not to any domain that is running inside a container. Of course the containers must be able to talk to another containers so it is possible to join a new container as a domain controller to a forest/domain. 

Instead of using VM's as mentioned above I would like to provision and deprovision containers (on the same host/vm) which will then be configured using DSC and the PowerShell module ActiveDirectoryDsc as domain controllers in a forest, child domain, trusted domain etc. 

Basically, I rather want to move all those VM's mentioned above that is today need to integration test the module into containers instead.

**Comment by vrapolinario** [comment]
Thanks for the details @johlju. Today we don't have that option available primarily because we treat containers as disposable units that can be scaled as needed. We heard this feedback from other customers already and it is something we're looking into, however, we don't have a timeline to provide at this moment.

**Comment by johlju** [comment]
> ... we treat containers as disposable units ...

That is what I'm looking for. I want the container(s) to live as long it takes for the integration tests to run, then it will be discarded. Next pipeline job will setup the container(s) to run the integration tests again. Today I would need to discard the VM's and rebuild them on next run.

**Comment by vrapolinario** [comment]
I understand the scenario you'd like to have AD for. The problem with implementing that today is that in a real AD environment, spinning up and down containers would leave lingering objects in the AD database, in addition to other authentication issues.

As I mentioned, this is something we're looking into as we have other customers asking for this. Unfortunately, I can't provide a timeline at this moment.

**Comment by vrapolinario** [comment]
I'll be closing this issue for now. Happy to continue the conversation and wanted to make sure customers do know this is something we'll investigate for future implementations. Unfortunately, we don't have a timeline to provide at this point.
